### ISSUE TITLE
Support multilabel classification and regression

### ISSUE THREAD

**Issue by j-adamczyk** [NannyML/nannyml#322]
**Motivation: describe the problem to be solved**

NannyML currently supports only single target classification and regression. Extending it to multilabel (multi-output) case would enable us to monitor our multilabel regression neural networks.

**Describe the solution you'd like**

Support for multilabel classifiers and regressors. Implementationally, this has a few possible approaches, but the simplest would be to treat this as a series of problems and average metrics for CBPE / DLE.

A nice addition would be to be able to configure how scores are combined. Average metric is the simplest one, but this treats all problems as equally important, and would result in no alarm if one metric is severely underperforming, but others are ok. Accepting a predefined option (mean, geometric mean, max) or Python callable, would be quite useful.

**Describe alternatives you've considered**

Create separate NannyML models for each target and monitor them separately. However, this has downsides:
- having to train and maintain multiple monitoring models separately
- having to aggregate the results ourselves, e.g. we have to look into 3 DLE results and compare them ourselves

**Additional context**

This is quite important for neural networks, since multilabel training can be used for expanding the training set. Even if some labels are not available, masked loss can be used. It also encourages generalization for representation learning, so it's quite useful. For tabular problems, Random Forest also supports this, as do scikit-multilearn models, with the same rationale.

**Comment by WojtekNML** [comment]
Hey Jakub,

Thanks for opening the issue. We've been considering adding multilabel classification (and to a lesser extent multilabel regression) for a while now, but we never had enough pull from users to actually prioritize it. Your request definitely makes us look at a roadmap again and at least consider prioritizing it. 

Would you be open to contributing these features? We will happily provide support and guidance. One example solution that should be reasonably easy to implement is extending our existing confusion matrix estimation feature to also estimate confusion matrix for multilabel problems. The implementation should be reasonably straightforward, as it's just doing binary classification multiple times and combining them as you outlined. Then extending it to all classification metrics should be very simple. 

Let me know if you'd like to contribute something along these lines!

**Comment by j-adamczyk** [comment]
I could try, this looks relatively straightforward. However, after delving into code details, I see overall a few things to discuss.

1. For probability calibration in CBPE, this will require training a separate calibration model per label. So an array of calibration models will need to be train and used for multilabel cases. This will be mostly exactly like the existing multiclass case, right?
2. Model predictions and actual targets are currently saved in Pandas DataFrames. However, columns with Numpy arrays (which would be result of multilabel classification) are not exactly well handled in Pandas from my experience. On a positive side, this would keep the code structure as-is (and quite readable), on the other will require weird handling in some cases. What do you think?
3. This would require changes in `cbpe.py` (add arguments and multilabel methods), and in `metrics.py` to add multilabel versions of metrics, and adding tests. Or did I forget something?

**Comment by WojtekNML** [comment]
Hey,

1. Yeah, it's basically like having multiple binary classification models. So not exactly like multiclass, but very similar.
2. That's above my pay grade. @nikml @nnansters help here pls. 
3. I think so. @nnansters can you confirm?

**Comment by nnansters** [comment]
Hey @j-adamczyk ,

Thanks for chipping in!

1.  Yup, sounds like multiclass. I tend to prefer using dictionaries for the use case of associating a label and a value. I like it more explicit and not depend just on array ordering.
2.  Yeah, that's a proper trade-off.  The idea of Numpy arrays in DataFrame columns does scare me somewhat. Turning the arrays into DataFrame columns seems like the cleanest way, but it might break the existing "single value target" implementation. Would have to delve a bit deeper for the impact that may have. But again, I do prefer making it more explicit; i.e. introducing multilabel support instead of hacking it in as embedded arrays. I will take a closer look it this as well.
3. Yes, that is very correct. Most of the work will be in the `metrics.py` module, but it should hopefully allow you to work in some isolation of the existing stuff (bar the changes for multilabel targets).
